src/Stratis.Bitcoin.Features.SmartContracts/Rules/ContractTransactionValidationRule.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using CSharpFunctionalExtensions;
+using NBitcoin;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Features.MemoryPool;
+using Stratis.SmartContracts.CLR;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.Rules
+{
+    public class ContractTransactionValidationRule : PartialValidationConsensusRule, ISmartContractMempoolRule
+    {
+        private readonly ICallDataSerializer callDataSerializer;
+
+        private readonly IList<IContractTransactionValidationLogic> internalRules;
+
+        public ContractTransactionValidationRule(ICallDataSerializer callDataSerializer, IList<IContractTransactionValidationLogic> internalRules)
+        {
+            this.callDataSerializer = callDataSerializer;
+            this.internalRules = internalRules;
+        }
+
+        public override Task RunAsync(RuleContext context)
+        {
+            Block block = context.ValidationContext.BlockToValidate;
+
+            foreach (Transaction transaction in block.Transactions.Where(x => !x.IsCoinBase && !x.IsCoinStake))
```

Comment:
Do we need the filter for CoinBase/CoinStake?

Reply:
Copied from SmartContractFormatRule but good call, I'll remove it.